packages/common/src/subchain/ReactSubchain.tsx

```diff
@@ -0,0 +1,123 @@
+import SubchainClient from "./SubchainClient";
+import { useContext, useEffect, useState, createContext } from "react";
+
+export function useCreateEdenChain(
+    edenAccount: string,
+    tokenAccount: string,
+    atomicAccount: string,
+    atomicmarketAccount: string,
+    wasmUrl: string,
+    stateUrl: string,
+    wsUrl: string,
+    slowMo: boolean
+): SubchainClient | null {
+    const [subchain, setSubchain] = useState<SubchainClient | null>(null);
+    useEffect(() => {
+        if (typeof window !== "undefined") {
+            let client: SubchainClient;
+            (async () => {
+                try {
+                    console.log("create SubchainClient");
+                    client = new SubchainClient();
+                    await client!.instantiateStreaming(
+                        edenAccount,
+                        tokenAccount,
+                        atomicAccount,
+                        atomicmarketAccount,
+                        fetch(wasmUrl),
+                        fetch(stateUrl),
+                        wsUrl,
+                        slowMo
+                    );
+                    setSubchain(client);
+                } catch (e) {
+                    console.error(e);
+                }
+            })();
+            return () => {
+                console.log("shutdown SubchainClient");
+                if (client) client.shutdown();
+            };
+        }
+    }, []);
+    return subchain;
+}
+
+export const EdenChainContext = createContext<SubchainClient | null>(null);
+
+export function useQuery<T = any>(query: string): T {
+    const client = useContext(EdenChainContext);
+    const [cachedQuery, setCachedQuery] = useState<string | null>();
+    // non-signalling state
+    const [state] = useState({
+        mounted: true,
+        cachedClient: null as SubchainClient | null,
+        subscribed: null as SubchainClient | null,
+        cachedQueryResult: null as any,
+    });
+    useEffect(() => {
+        return () => {
+            state.mounted = false;
+        };
+    }, []);
+    useEffect(() => {
+        if (client && state.subscribed !== client) {
+            state.subscribed = client;
+            client.notifications.push((c) => {
+                if (state.mounted && c === state.subscribed) {
+                    setCachedQuery(null);
+                    state.subscribed = null;
+                }
+            });
+        }
+    });
+    if (state.cachedClient !== client || query !== cachedQuery) {
+        state.cachedClient = client;
+        setCachedQuery(query);
+        if (client?.subchain) {
+            state.cachedQueryResult = client.subchain.query(query);
+        } else {
+            state.cachedQueryResult = null;
+        }
+    }
+    return state.cachedQueryResult;
+}
```

Comment:
it's the first time that I'm seeing a `useState` variable being updated without the set functions. Are these valid operations? Is it not presenting any weirdness? 

Why not follow the common pattern as in the below example?
```ts
const [state, setState] = useState(...)
// ...
setState({...state, mounted: false})

// ...
setState({...state, subscribed: client})
```

Reply:
`setState` schedules a render. Modifying the object directly doesn't. I marked it `non-signalling state`